### PR TITLE
[IMP] runbot: add kill button

### PR DESCRIPTION
**Pull request by Icallhimtest**
Allow logged users to kill builds that are pending or testing.

The kill button solves the following problem:

When you have a community and enterprise branch with pull request.
The enterprise PR gets fetched before the branch by runbot, so it
builds with the community master branch instead of the corresponding
community branch. As the branch shares the same HEAD as the PR,
so no other build is done for the branch.

Previously you had to wait the build to finish building, then rebuild it
from the branch, taking twice the time necessary. Now you'll be able to
kill & rebuild it just after pushing the branch.

Also having a kill button is globally a nice to have feature.

### CODE REVIEW

File: runbot/runbot.py [hunk 1]
```diff
@@ -1541,10 +1544,20 @@ def build(self, build_id=None, search=None, **post):
...
+    def build_force(self, build_id, search=None, **post):
+        registry, cr, uid = request.registry, request.cr, request.uid
         repo_id = registry['runbot.build']._force(cr, uid, [int(build_id)])
-        return werkzeug.utils.redirect('/runbot/repo/%s' % repo_id)
+        return werkzeug.utils.redirect('/runbot/repo/%s' % repo_id + ('?search=%s' % search if search else ''))
+
+    @http.route(['/runbot/build/<build_id>/kill'], type='http', auth="public", methods=['POST'], csrf=False)
+    def build_kill(self, build_id, search=None, **post):
+        registry, cr = request.registry, request.cr
+        build = registry['runbot.build'].browse(cr, SUPERUSER_ID, build_id)
```
Comment (Icallhimtest) [inline runbot/runbot.py]: (used SUPERUSER_ID because I don't believe users have rights to write on builds...)

File: runbot/runbot.py [hunk 2]
```diff
@@ -1555,7 +1568,7 @@ def badge(self, repo_id, branch, theme='default'):
         domain = [('repo_id', '=', repo_id),
                   ('branch_id.branch_name', '=', branch),
                   ('branch_id.sticky', '=', True),
-                  ('state', 'in', ['testing', 'running', 'done']),
+                  ('state', 'in', ['testing', 'running', 'done']), # don't know what this route does, need to add 'deathrow' ?
```
Comment (Icallhimtest) [inline runbot/runbot.py]: need input here

File: runbot/runbot.py [hunk 3]
```diff
@@ -1541,10 +1544,20 @@ def build(self, build_id=None, search=None, **post):
         return request.render("runbot.build", context)
 
     @http.route(['/runbot/build/<build_id>/force'], type='http', auth="public", methods=['POST'], csrf=False)
-    def build_force(self, build_id, **post):
-        registry, cr, uid, context = request.registry, request.cr, request.uid, request.context
+    def build_force(self, build_id, search=None, **post):
+        registry, cr, uid = request.registry, request.cr, request.uid
         repo_id = registry['runbot.build']._force(cr, uid, [int(build_id)])
-        return werkzeug.utils.redirect('/runbot/repo/%s' % repo_id)
+        return werkzeug.utils.redirect('/runbot/repo/%s' % repo_id + ('?search=%s' % search if search else ''))
```
Comment (KangOl) [inline runbot/runbot.py]: Should be another commit

File: runbot/runbot.xml [hunk 4]
```diff
@@ -254,6 +255,7 @@
     </template>
 
     <template id="runbot.build_name">
+        <t t-if="bu['state']=='deathrow'"><i class="text-info fa fa-times-circle"/> killing</t>
```
Comment (KangOl) [inline runbot/runbot.xml]: I propose `fa-crosshairs`
http://fontawesome.io/icon/crosshairs/

File: runbot/runbot.xml [hunk 5]
```diff
@@ -284,9 +286,12 @@
                         <li><a t-attf-href="http://{{bu['domain']}}/?db={{bu['real_dest']}}-base">Connect base <i class="fa fa-sign-in"></i></a></li>
                         <li><a t-attf-href="http://{{bu['domain']}}/">Connect <i class="fa fa-sign-in"></i></a></li>
                     </t>
-                    <li t-if="bu['state'] in ['done','running'] and bu_index==0" groups="runbot.group_user">
+                    <li t-if="bu['state'] in ['done','running', 'deathrow'] and bu_index==0" groups="runbot.group_user">
                         <a href="#" class="runbot-rebuild" t-att-data-runbot-build="bu['id']">Rebuild <i class="fa fa-refresh"/></a>
                     </li>
+                    <li t-if="bu['state'] in ['pending','testing']" groups="runbot.group_user">
+                        <a href="#" class="runbot-kill" t-att-data-runbot-build="bu['id']">Kill <i class="fa fa-times-circle"/></a>
```
Comment (KangOl) [inline runbot/runbot.xml]: Ditto

File: runbot/runbot.py (hunk 1 above)
Comment (KangOl) [inline runbot/runbot.py]: This route should be `auth=user`, then `uid` may be used.

File: runbot/runbot.py [hunk 6]
```diff
@@ -1126,7 +1126,10 @@ def _schedule(self, cr, uid, ids, context=None):
         default_timeout = int(icp.get_param(cr, uid, 'runbot.timeout', default=1800)) / 60
 
         for build in self.browse(cr, uid, ids, context=context):
-            if build.state == 'pending':
+            if build.state == 'deathrow':
+                build._kill(result='killed')
```
Comment (KangOl) [inline runbot/runbot.py]: Should use another `result` value to differenciate build that have been manually killed from one killed because of timeout.

File: runbot/runbot.py (hunk 2 above)
Comment (KangOl) [inline runbot/runbot.py]: This route return the image for the badge shown in README of odoo/odoo [1]
You need to ignore build manually killed (next line)

[1] https://github.com/odoo/odoo/#readme

File: runbot/runbot.py (hunk 6 above)
Comment (Icallhimtest) [inline runbot/runbot.py]: didn't want to create a new result type, but I guess you're right, I'll change it.

File: runbot/runbot.py (hunk 1 above)
Comment (Icallhimtest) [inline runbot/runbot.py]: shall I adapt `/runbot/build/<build_id>/force`'s route too ? Not really sure why we're not using auth=user there too.

Still, group_user should be able to kill builds, but they do not have the access rights to do so, so SUPERUSER_ID will need to be user, no?

**Comment by Icallhimtest** [comment]
Woo !

**Comment by KangOl** [comment]
Still, there was a few errors: https://github.com/odoo/odoo-extra/compare/37ee888bbf6378cce62ae74bcf5ab8674cdff8e5...8799edeab1f5a6bf678e1702739630b782e9f89c

**Comment by Icallhimtest** [comment]
indeed...

Thanks
